k-distribution/src/main/scripts/lib/pyk/pyk/kcfg.py

```diff
@@ -1,59 +1,100 @@
 import json
-import threading
-from typing import Final
+from abc import ABC
+from dataclasses import dataclass
+from itertools import chain
+from threading import RLock
+from typing import (
+    Any,
+    Dict,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Union,
+)
 
 from graphviz import Digraph
 
-from .cli_utils import fatal, notif
-from .kast import KClaim, KFlatModule, KImport, KInner, KRule
+from .kast import TRUE, KClaim, KInner, KRule
 from .kastManip import (
     buildRule,
+    compose_substs,
     matchWithConstraint,
     minimizeSubst,
     mlAnd,
     simplifyBool,
-    substitute,
-    substToMap,
     unsafeMlPredToBool,
 )
-from .utils import compare_short_hashes, hash_str, shorten_hashes
+from .ktool import KPrint
+from .utils import compare_short_hashes, shorten_hashes
 
 
 class KCFG:
-    _FIELDS: Final = (
-        'states',
-        'graph',
-        'abstractions',
-        'loop',
-        'split',
-        'init',
-        'target',
-        'terminal',
-        'stuck',
-        'subsumptions',
-        'frontier',
-    )
-
-    _NODE_ATTRS: Final = ('init', 'target', 'stuck', 'terminal', 'frontier', 'loop', 'split')
+
+    @dataclass(frozen=True)
+    class Node:
+        id: str
+        term: KInner
+
+        def __init__(self, term: KInner):
+            object.__setattr__(self, 'id', term.to_hash())
+            object.__setattr__(self, 'term', term)
+
+        def to_dict(self) -> Dict[str, Any]:
+            return {'id': self.id, 'term': self.term.to_dict()}
+
+    class EdgeLike(ABC):
+        source: 'KCFG.Node'
+        target: 'KCFG.Node'
+
+    @dataclass(frozen=True)
+    class Edge(EdgeLike):
+        source: 'KCFG.Node'
+        target: 'KCFG.Node'
```

Comment:
Why are we repeating these fields from the superclass `EdgeLike`? Is this some Python convention I'm not aware of, or actually a requirement?

Reply:
Good point, in general we could omit these type hints in the subclasses. However, in this case they are necessary for the `@dataclass` decorator to work properly.